zerver/tests/test_slack_importer.py

```diff
@@ -506,26 +514,51 @@ def test_convert_slack_workspace_messages(self, mock_get_total_messages_and_user
         self.assertEqual(message_json['zerver_message'], zerver_message1 + zerver_message2)
         self.assertEqual(message_json['zerver_usermessage'], zerver_usermessage1 + zerver_usermessage2)
 
-    @mock.patch("zerver.lib.slack_data_to_zulip_data.build_avatar_url")
-    @mock.patch("zerver.lib.slack_data_to_zulip_data.build_avatar", return_value = [])
+    @mock.patch('requests.get', side_effect=mocked_requests_get)
+    @mock.patch("zerver.lib.slack_data_to_zulip_data.build_avatar")
+    @mock.patch("zerver.lib.slack_data_to_zulip_data.allocate_ids")
     @mock.patch("zerver.lib.slack_data_to_zulip_data.get_user_data")
     def test_slack_import_to_existing_database(self, mock_get_user_data: mock.Mock,
-                                               mock_build_avatar_url: mock.Mock,
-                                               mock_build_avatar: mock.Mock) -> None:
+                                               mock_allocate_ids: mock.Mock,
+                                               mock_build_avatar: mock.Mock,
+                                               mock_reuqests: mock.Mock) -> None:
         test_slack_zip_file = os.path.join(settings.DEPLOY_ROOT, "zerver", "fixtures",
                                            "slack_fixtures", "test_slack_importer.zip")
         test_realm_subdomain = 'test-slack-import'
         output_dir = '/tmp/test-slack-importer-data'
+        realm_id = 10000
+        user_id = 100001
         token = 'valid-token'
+        avatar_data = [{'s3_path': '',
+                        'realm_id': realm_id,
+                        'user_profile_id': user_id,
+                        'path': 'https://ca.slack-edge.com/T5YFFM2QY-U6006P1CN-gd41c3c33cbe'}]
 
         user_data_fixture = os.path.join(settings.DEPLOY_ROOT, "zerver", "fixtures",
                                          "slack_fixtures", "user_data.json")
         mock_get_user_data.return_value = ujson.load(open(user_data_fixture))['members']
+        mock_build_avatar.return_value = avatar_data
+
+        # Side effect to return Realm id as the value 'realm_id'
+        # arg2 is added corresponding to the data of that object in /fixtures/slack_fixtures
+        def allocate_ids_side_effect(arg1: Any, arg2: int) -> List[int]:
+            if arg1 == Realm and arg2 == 1:
+                return [realm_id]
+            elif arg1 == UserProfile and arg2 == 5:
+                return [100001, 100002, 100003, 100004, 100005]
+            else:
+                return allocate_ids(arg1, arg2)
+        mock_allocate_ids.side_effect = allocate_ids_side_effect
```

Comment:
While this works, I think a cleaner approach would be to just do a database query to look up what realm ID was used for the realm being tested (e.g. `get_realm("subdomain").id, etc.)

Reply:
@timabbott The problem here would be that the `avatar_data` [here](https://github.com/rheaparekh/zulip/blob/6346ce3c025f96accf5dade40a9fd8f53574190c/zerver/tests/test_slack_importer.py#L532-L535)  would be used in the import script [here](https://github.com/zulip/zulip/blob/master/zerver/lib/export.py#L1354) and we need the `realm_id` for it. Hence we can't get the `realm_id` from `get_realm` beforehand. (We have to make sure `avatar_data` IDs are consistent with the import IDs before the import is done.)
 One thing we can do here is mock the function [import_uploads_s3](https://github.com/zulip/zulip/blob/master/zerver/lib/export.py#L1339), but that would be against the purpose of testing the uploads.

Reply:
To give a better explanation, I have this case: We have to give a `user_profile_id` in the `avatar_data`, which is used during the import. Also when the slack data conversion script runs, we get list of the user IDs from `allocate_ids`. So if the avatar_data `user_profile_id` isn't present in this list, we will get an error while running the import at this [line](https://github.com/zulip/zulip/blob/master/zerver/lib/export.py#L1354).

I have changed it now to a better implementation. I am using allocate ids initially to get the ID lists for `Realm` and `UserProfile`(the mock function will return these), and then passing it to the `avatar_data`.